processing/src/main/java/org/apache/druid/math/expr/ExpressionTypeConversion.java

```diff
@@ -55,6 +55,9 @@ public static ExpressionType autoDetect(ExprEval eval, ExprEval otherEval)
   {
     ExpressionType type = eval.type();
     ExpressionType otherType = otherEval.type();
+    if (type == otherType && eval.value() != null && otherEval.value() != null) {
```

Comment:
why does this need more than `type == otherType`? Later on we do
```
    type = eval.value() != null ? type : otherType;
    otherType = otherEval.value() != null ? otherType : type;
```
so if the value is null then it just takes on the other type, so i think being null or not here doesn't matter, because if the value is null and the same type then it is fine

Reply:
sure - I was too cautious...if it will be `null` ; the type will be `String` as per the comment above the method - so it will be only `==` if the other is also a `String`
```
In this mode, null values are {@link ExprType#STRING} typed, despite
potentially coming from an underlying numeric column, or when an underlying column was completely missing and so all values are null
```

regarding those lines:
if `eval.value() == null`  ; after the 1st line `type = otherType` ; independently from the the 2nd line that will mean that effectively `numeric(otherType, otherType)` will be called...
```
if(eval.value() == null) {
  return otherType;
}
if(otherEval.value() == null) {
  return type;
}
return numeric(type, otherType);
```

which made me wonder... can `ExpressionType.getType() == COMPLEX`  for this method? if it could this change could alter its behaviour - as earlier `DOUBLE` was returned in that case


to avoid changing that behaviour - for now I've changed it to:
```java
    if (type == otherType && type.getType().isPrimitive()) {
      return type;
    }
```
should these pass / fail / or shouldn't exists at all?
```java
    final ExprEval<?> complexEval = ExprEval.ofComplex(ExpressionType.UNKNOWN_COMPLEX, new Object());
    Assert.assertEquals(ExpressionType.DOUBLE, ExpressionTypeConversion.autoDetect(longEval, complexEval));
    Assert.assertEquals(ExpressionType.DOUBLE, ExpressionTypeConversion.autoDetect(doubleEval, complexEval));
    Assert.assertEquals(ExpressionType.DOUBLE, ExpressionTypeConversion.autoDetect(arrayEval, complexEval));
    Assert.assertEquals(ExpressionType.DOUBLE, ExpressionTypeConversion.autoDetect(complexEval, complexEval));
```

Reply:
good question, so this function predates expressions supporting complex types, and shouldn't be used by anything that handles arrays or complex types.

additionally, the comment about nulls always being strings isn't quite as true as it once was, null values coming from columns use the correctly typed null constants, so its mostly null constants and nulls created without additional type information (bestEffortOf) which are string types.

I guess autoDetect should be an invalid exception if it is trying to process complex types, but autoDetect is also only used inside of eval methods as far as I can tell, `Expr.getOutputType` all use other methods in this class like `function` and `operator` which do handle complex and array types and ensure that they are compatible, so I'm not sure it proves to be much of a problem in practice.